### PR TITLE
:sparkle: Add defaulting and validation webhook

### PR DESCRIPTION
**Pull request by swapnilgm**
Signed-off-by: Swapnil Mhamane <swapnil.mhamane@sap.com>

**What this PR does / why we need it**:
This PR add the webhook for defaulting and validating etcd CRD

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
TODO:
- [ ] Local development setup
- [ ] Test

**Release note**:

```action operator
Druid now host the webhook for validation and defaulting of etcd crd. Oerator should apply the webhook configuration manifest from `config/webhook` with appropriate changes to enable webhook.
```

### CODE REVIEW

File: api/v1alpha1/etcd_webhook.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+			panic(result)
+		}
+	}
+
+	if r.Spec.VolumeClaimTemplate == nil {
+		r.Spec.VolumeClaimTemplate = new(string)
+		*r.Spec.VolumeClaimTemplate = r.Name
+	}
+	if r.Spec.StorageCapacity == nil {
+		r.Spec.StorageCapacity = resource.NewScaledQuantity(20, resource.Giga)
```
Comment (amshuman-kr) [inline api/v1alpha1/etcd_webhook.go:125]: Constant?

File: api/v1alpha1/etcd_webhook.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	// TODO: should we(is it feasible to) import etcd-backup-restore for such defaults?
+	if r.Spec.Etcd.DefragmentationSchedule == nil {
+		defaultDefragmentationSchedule := "0 0 */3 * *" //Per 3 day
```
Comment (amshuman-kr) [inline api/v1alpha1/etcd_webhook.go:130]: Constant?

File: api/v1alpha1/etcd_webhook.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+		r.Spec.Etcd.DefragmentationSchedule = &defaultDefragmentationSchedule
+	}
+	r.defaultBackupSpec()
+
+}
+
+func (r *Etcd) defaultBackupSpec() {
```
Comment (amshuman-kr) [inline api/v1alpha1/etcd_webhook.go:137]: Constants?

File: api/v1alpha1/etcd_webhook.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+		return nil
+	}
+
+	return errors.NewInvalid(schema.GroupKind{Group: GroupVersion.Group, Kind: "Etcd"}, r.Name, allErrs)
+}
+
+func (r *Etcd) validateEtcdUpdate(old *Etcd) error {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, validation.ValidateImmutableField(r.Spec.Backup.Store, r.Spec.Backup.Store, field.NewPath("spec", "backup", "store"))...)
```
Comment (amshuman-kr) [inline api/v1alpha1/etcd_webhook.go:221]:
```suggestion
	allErrs = append(allErrs, validation.ValidateImmutableField(r.Spec.Backup.Store, old.Spec.Backup.Store, field.NewPath("spec", "backup", "store"))...)
```

File: api/v1alpha1/etcd_webhook.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+			supportedProviderList := []string{"aws", "azure", "gcp", "alicloud", "openstack", "Local", "S3", "ABS", "GCS", "OSS", "Swift"}
+			found := false
+			for _, val := range supportedProviderList {
+				if val == string(*r.Spec.Backup.Store.Provider) {
+					found = true
+					break
+				}
+			}
+			if found {
```
Comment (amshuman-kr) [inline api/v1alpha1/etcd_webhook.go:255]:
```suggestion
			if !found {
```

File: controllers/etcd_controller.go [hunk 6]
```diff
@@ -814,7 +790,7 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
 	}
 
 	var statefulsetReplicas int
-	if etcd.Spec.Replicas != 0 {
+	if *etcd.Spec.Replicas != 0 {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go:793]: Nil check?

**Comment by swapnilgm** [comment]
Closing this PR. Since the branch has been moved from swapnilgm/etcd-duird/feature/webhook -> gardener/etcd-druid/feature/webhook

**Pull request closed without merging**